File: src/models/Player.ts
import { Card } from './Card';

/**
 * Player class representing a game player with hand, captured cards, and score
 */
export class Player {
  name: string;
  hand: Card[];
  captured: Card[];  // All captured cards
  mulles: Card[];    // Cards representing mulle points (one per pair)
  tabbe: number;     // Number of captures from empty board

  constructor(name: string) {
    this.name = name;
    this.hand = [];
    this.captured = [];
    this.mulles = [];
    this.tabbe = 0;
  }

  /**
   * Add cards to hand
   */
  addToHand(cards: Card[]): void {
    this.hand.push(...cards);
  }

  /**
   * Remove a card from hand
   */
  removeFromHand(card: Card): void {
    const index = this.hand.indexOf(card);
    if (index > -1) {
      this.hand.splice(index, 1);
    }
  }

  /**
   * Record a mulle (one card per pair)
   */
  recordMulle(card: Card): void {
    this.mulles.push(card);
  }

  /**
   * Record captured cards
   */
  recordCapture(cards: Card[]): void {
    this.captured.push(...cards);
  }

  /**
   * Calculate total mulle points
   */
  totalMullePoints(): number {
    let points = 0;
    for (const card of this.mulles) {
      if (card.rank === 'A') {
        points += 14;
      } else if (['J', 'Q', 'K'].includes(card.rank)) {
        points += card.valueOnBoard();
      } else {
        points += parseInt(card.rank, 10);
      }
    }
    return points;
  }

  /**
   * String representation
   */
  toString(): string {
    return this.name;
  }
}

File: archive/legacy/mulle/models/player.py
from typing import List
from .card import Card

class Player:
    def __init__(self, name: str):
        self.name = name
        self.hand: List[Card] = []
        self.captured: List[Card] = []  # non-mulle captured
        self.mulles: List[Card] = []     # cards representing mulle points
        self.tabbe: int = 0

    def add_to_hand(self, cards: List[Card]):
        self.hand.extend(cards)

    def remove_from_hand(self, card: Card):
        self.hand.remove(card)

    def record_mulle(self, card: Card):
        self.mulles.append(card)

    def record_capture(self, cards: List[Card]):
        self.captured.extend(cards)

    def total_mulle_points(self) -> int:
        points = 0
        for c in self.mulles:
            if c.rank == "A":
                points += 14  # assumption
            elif c.rank in ["J", "Q", "K"]:
                # Face cards map to board values anyway
                points += c.value_on_board()
            else:
                points += int(c.rank)
        return points

    def __str__(self):
        return self.name

